service/history/workflow/update/update.go

```diff
@@ -482,11 +494,24 @@ func (u *Update) onAcceptanceMsg(
 			return
 		}
 		u.request = nil
+
+		// If Update is accepted and completed on the same WFT, then its state transitions
+		// from ProvisionallyAccepted to ProvisionallyCompleted and, on commit of acceptance
+		// message, Update is in ProvisionallyCompleted state.
+		// To prevent race condition with checks in WaitLifecycleStage, accepted future
+		// must be set after outcome future.
+		// Therefore, if Update is in ProvisionallyCompleted state, it skips Accepted state
+		// and setting of accepted future is delayed. Update goes to ProvisionallyCompletedAfterAccepted
+		// state instead, which is handled in onResponseMsg.
```

Comment:
```suggestion
		// If the Update is accepted *and* completed in the same WFT, then its state has transitioned
		// from ProvisionallyAccepted to ProvisionallyCompleted in onResponseMsg by the
		// time we get here.
		// 
		// Now, to prevent a race condition in WaitLifecycleStage, the accepted future
		// cannot be set here right now, as it must be set *after* the outcome future.
		// 
		// So instead, the state is set to ProvisionallyCompletedAfterAccepted here,  
		// and onResponseMsg's OnAfterCommit callback will set the futures in the correct order.
```